### PR TITLE
[WIP] added function for syntetic sideslip, fusion deactivated

### PR DESCRIPTION
**Pull request by CarlOlsson**
This can be merged when:
- [ ] Logic is implemented when to fuse syntetic sideslip
- [ ] Functionality verified with test flights and RPL logs 
- [ ] Code reviewed

### CODE REVIEW

File: EKF/sideslip_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		// update covariance matrix via Pnew = (I - KH)P = P - KHP
+		for (unsigned row = 0; row < _k_num_states; row++) {
+			for (unsigned column = 0; column < _k_num_states; column++) { // Here it will be a lot of zeros, should optimize that...
+				KH[row][column] = Kfusion[row] * H_BETA[column];
+			}
+		}
+
+		for (unsigned row = 0; row < _k_num_states; row++) {
+			for (unsigned column = 0; column < _k_num_states; column++) {
+				for (unsigned i = 0; i < _k_num_states; i++) { // Check if this is correct matrix multiplication!
```
Comment (priseborough) [inline EKF/sideslip_fusion.cpp]: Resolve

File: EKF/sideslip_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+        // Syntetic sideslip measurement sample has passed check so record it
+        _time_last_beta_fuse = _time_last_imu;
+
+		// Fuse syntetic sideslip measurement
+		fuse(Kfusion, _beta_innov); //Why calculate angle error when it is always zero?
+
+		// update covariance matrix via Pnew = (I - KH)P = P - KHP
+		for (unsigned row = 0; row < _k_num_states; row++) {
+			for (unsigned column = 0; column < _k_num_states; column++) { // Here it will be a lot of zeros, should optimize that...
```
Comment (priseborough) [inline EKF/sideslip_fusion.cpp]: Resolve

File: EKF/sideslip_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        // Compute the ratio of innovation to gate size
+		_beta_test_ratio = sq(_beta_innov) / (sq(fmaxf(_params.beta_innov_gate, 1.0f)) * _beta_innov_var);
+
+		// if the innovation consistency check fails then don't fuse the sample and indicate bad beta health
+		if (_beta_test_ratio > 1.0f) {
+			_beta_healthy = false;
+			return;
+		}
+
+        // Syntetic sideslip measurement sample has passed check so record it
```
Comment (priseborough) [inline EKF/sideslip_fusion.cpp]: 'Synthetic' is the correct spelling

File: EKF/sideslip_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		}
+
+		for (unsigned row = 0; row < _k_num_states; row++) {
+			for (unsigned column = 0; column < _k_num_states; column++) {
+				for (unsigned i = 0; i < _k_num_states; i++) { // Check if this is correct matrix multiplication!
+					KHP[row][column] += KH[row][i] * P[i][column];
+				}
+			}
+		}
+
```
Comment (priseborough) [inline EKF/sideslip_fusion.cpp]: Add the check used by other fusion operations where the state update and covariance update is not performed and a fault is recorded if any diagonals of KHP are greater than the corresponding diagonals on P.

File: EKF/sideslip_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
...*(SH_BETA[12] + 2.0f*q1*q2)) + P[0][3]*(SH_BETA[5]*SH_BETA[8] - SH_BETA[1]*SH_BETA[4]*SH_BETA[9]) + P[1][3]*(SH_BETA[5]*SH_BETA[10] - SH_BETA[1]*SH_BETA[4]*SH_BETA[11]) + P[2][3]*(SH_BETA[5]*SH_BETA[11] + SH_BETA[1]*SH_BETA[4]*SH_BETA[10]) - P[3][3]*(SH_BETA[5]*SH_BETA[9] + SH_BETA[1]*SH_BETA[4]*SH_BETA[8]) + P[6][3]*(SH_BETA[5]*(2.0f*q0*q1 + 2.0f*q2*q3) + SH_BETA[1]*SH_BETA[4]*(2.0f*q0*q2 - 2.0f*q1*q3))) + (SH_BETA[5]*(2.0f*q0*q1 + 2.0f*q2*q3) + SH_BETA[1]*SH_BETA[4]*(2.0f*q0*q2 - 2.0f*q1*q3))*(P[22][6]*(SH_BETA[5]*(SH_BETA[12] - 2.0f*q1*q2) + SH_BETA[1]*SH_BETA[4]*SH_BETA[7]) - P[4][6]*(SH_BETA[5]*(SH_BETA[12] - 2.0f*q1*q2) + SH_BETA[1]*SH_BETA[4]*SH_BETA[7]) + P[5][6]*(SH_BETA[6] - SH_BETA[1]*SH_BETA[4]*(SH_BETA[12] + 2.0f*q1*q2)) - P[23][6]*(SH_BETA[6] - SH_BETA[1]*SH_BETA[4]*(SH_BETA[12] + 2.0f*q1*q2)) + P[0][6]*(SH_BETA[5]*SH_BETA[8] - SH_BETA[1]*SH_BETA[4]*SH_BETA[9]) + P[1][6]*(SH_BETA[5]*SH_BETA[10] - SH_BETA[1]*SH_BETA[4]*SH_BETA[11]) + P[2][6]*(SH_BETA[5]*SH_BETA[11] + SH_BETA[1]*SH_BETA[4]*SH_BETA[10]) - P[3][6]*(SH_BETA[5]*SH_BETA[9] + SH_BETA[1]*SH_BETA[4]*SH_BETA[8]) + P[6][6]*(SH_BETA[5]*(2.0f*q0*q1 + 2.0f*q2*q3) + SH_BETA[1]*SH_BETA[4]*(2.0f*q0*q2 - 2.0f*q1*q3))));
```
Comment (priseborough) [inline EKF/sideslip_fusion.cpp]: Duplicate code. We have already calculated SH_BETA above.

File: EKF/common.h [hunk 6]
```diff
@@ -319,6 +324,11 @@ struct parameters {
 		tas_innov_gate = 5.0f;		
   		eas_noise = 1.4f;			
 
+  		// sysntetic sideslip fusion
+  		beta_innov_gate = 5.0f;		// [CHECK THIS VALUE]
```
Comment (priseborough) [inline EKF/common.h]: A description of the parameter including units is required here, not a reminder note

File: EKF/common.h [hunk 7]
```diff
@@ -319,6 +324,11 @@ struct parameters {
...
+  		beta_noise = 0.03f; // Check this value
```
Comment (priseborough) [inline EKF/common.h]: A description of the parameter including units is required here, not a reminder note.

File: EKF/common.h [hunk 8]
```diff
@@ -225,6 +225,11 @@ struct parameters {
 	float tas_innov_gate;		// True Airspeed Innovation consistency gate size in standard deciation
   	float eas_noise;			// EAS measurement noise standard deviation used for airspeed fusion [m/s]
 
+  	// syntetic sideslip fusion
+  	float beta_innov_gate;		// Syntetic sideslip innovation consistency gate size in standard deviation
+  	float beta_noise;			// Syntetic sideslip noise
```
Comment (priseborough) [inline EKF/common.h]: Please indicate units - eg (rad)

File: EKF/common.h [hunk 9]
```diff
@@ -225,6 +225,11 @@ struct parameters {
 	float tas_innov_gate;		// True Airspeed Innovation consistency gate size in standard deciation
   	float eas_noise;			// EAS measurement noise standard deviation used for airspeed fusion [m/s]
 
+  	// syntetic sideslip fusion
+  	float beta_innov_gate;		// Syntetic sideslip innovation consistency gate size in standard deviation
```
Comment (priseborough) [inline EKF/common.h]: 'Synthetic' spelling. Also please use same convention as other parameters and append units in parenthesis, eg (STD)

File: EKF/airspeed_fusion.cpp [hunk 10]
```diff
@@ -136,7 +136,7 @@ void Ekf::fuseAirspeed()
 		// Compute the ratio of innovation to gate size
 		_tas_test_ratio = sq(_airspeed_innov) / (sq(fmaxf(_params.tas_innov_gate, 1.0f)) * _airspeed_innov_var);
 
-		// If the innovation consistency check fails then don't fuse the sample and indicate bad airspeed health
+		// if the innocation consistency check fails then don't fuse the sample and indicate bad airspeed health
```
Comment (priseborough) [inline EKF/airspeed_fusion.cpp]: 'innovation'

File: EKF/control.cpp [hunk 11]
```diff
@@ -407,7 +407,7 @@ void Ekf::controlHeightSensorTimeouts()
 				_control_status.flags.ev_hgt = false;
 				// request a reset
 				reset_height = true;
-				ECL_INFO("EKF gps hgt timeout - reset to GPS");
+				printf("EKF gps hgt timeout - reset to GPS\n");
```
Comment (priseborough) [inline EKF/control.cpp]: replace with ECL_INFO

File: EKF/ekf.h [hunk 12]
```diff
@@ -180,6 +186,7 @@ class Ekf : public EstimatorInterface
 	bool _fuse_vert_vel;		// gps vertical velocity measurement should be fused
 	bool _fuse_flow;		// flow measurement should be fused
 	bool _fuse_hagl_data;		// if true then range data will be fused to estimate terrain height
+	bool _fuse_beta;		// syntetic sideslip measurement should be fused
```
Comment (priseborough) [inline EKF/ekf.h]: 'synthetic'

File: EKF/ekf.h [hunk 13]
```diff
@@ -209,6 +217,9 @@ class Ekf : public EstimatorInterface
 	float _airspeed_innov;		// airspeed measurement innovation
 	float _airspeed_innov_var;	// airspeed measurement innovation variance
 
+	float _beta_innov;		// syntetic sideslip measurement innovation
```
Comment (priseborough) [inline EKF/ekf.h]: 'synthetic'

File: EKF/ekf.h [hunk 14]
```diff
@@ -209,6 +217,9 @@ class Ekf : public EstimatorInterface
...
+	float _beta_innov_var;	// syntetic sideslip measurement innovation variance
```
Comment (priseborough) [inline EKF/ekf.h]: 'synthetic'

File: EKF/ekf.h [hunk 15]
```diff
@@ -300,6 +311,9 @@ class Ekf : public EstimatorInterface
 	// fuse airspeed measurement
 	void fuseAirspeed();
 
+	// fuse sideslip measurement
```
Comment (priseborough) [inline EKF/ekf.h]: Make it clear in the comment that this is not a real measurement, eg 'fuse synthetic zero sideslip measurement'

File: EKF/estimator_interface.h [hunk 16]
```diff
@@ -63,9 +63,12 @@ class EstimatorInterface
 	// gets the innovations of the earth magnetic field measurements
 	virtual void get_mag_innov(float mag_innov[3]) = 0;
 
-	// gets the innovation of airspeed measurement
+	// gets the innovation of the airspeed measurement
  	virtual void get_airspeed_innov(float *airspeed_innov) = 0;
 
+	// gets the innovation of the syntetic sideslip measurement
```
Comment (priseborough) [inline EKF/estimator_interface.h]: 'synthetic'

File: EKF/estimator_interface.h [hunk 17]
```diff
@@ -276,14 +282,14 @@ class EstimatorInterface
 	float _gps_origin_eph; // horizontal position uncertainty of the GPS origin
 	float _gps_origin_epv; // vertical position uncertainty of the GPS origin
 	struct map_projection_reference_s _pos_ref;    // Contains WGS-84 position latitude and longitude (radians)
-
 	// innovation consistency check monitoring ratios
-	float _yaw_test_ratio;          // yaw innovation consistency check ratio
+		float _yaw_test_ratio;          // yaw innovation consistency check ratio
```
Comment (priseborough) [inline EKF/estimator_interface.h]: fix indenting

**Comment by CarlOlsson** [comment]
this is now outdated

**Pull request closed without merging**